### DISCUSSION TITLE
[DrainingHash] pulsar-client 4.0.1 with java.lang.ClassNotFoundException

### DISCUSSION THREAD

**Discussion by unicojoyhug** [apache/pulsar#23744] [Q&A]
Hi, 

### Issue:
Upgrading `org.apache.pulsar:pulsar-client from version` from `3.3.1`  to `4.0.1` - it fails to create the client with `ClassNotFoundException` as below:

```
Failed to instantiate [org.apache.pulsar.client.api.PulsarClient]: 
Factory method 'pulsarClient' threw exception with message: org/apache/pulsar/common/policies/data/DrainingHash
```

### Implementation:
I use `pulsar-client` to create a client to consume events with adding `AuthenticationOAuth2`.
- `privateKey` is added with `data:application/json;base64,<base64-encoded value>` following [doc - authentication-types](https://pulsar.apache.org/docs/4.0.x/security-oauth2/#authentication-types)

### What I checked:
- checked the source package where it referenced `DrainingHashImpl` :  the import path doesn't contain `DrainingHash`
- checked where it was found in pulsar repo: `DrainingHash` interface is found in [pulsar-client-admin-api/common](https://github.com/apache/pulsar/blob/master/pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/DrainingHash.java)
  - maybe it is related to this change? ref: [PR - [improve][broker][PIP-379] Add observability stats for "draining hashes"](https://github.com/apache/pulsar/pull/23429)

### What I have tried:
- tried to use `pulsar-client-all` instead to see if I can get these all included, but it failed to find the `DrainingHash` interface for `DrainingHashImpl`

Let me know if I need to supply more information or write this to somewhere else.
I'd love to know what I should do to avoid this error.

**Comment by lhotari** [comment]
You might have multiple versions of the Pulsar client in the classpath. The DrainingHash interface is included in the pulsar-client-admin-api module. One way to resolve this problem is to use the [Pulsar BOM (unless you are using Spring Boot)](https://pulsar.apache.org/docs/4.0.x/client-libraries-java-setup/#pulsar-bom). For Spring Boot there's a [specific solution](https://pulsar.apache.org/docs/4.0.x/client-libraries-java-setup/#spring-boot).
